src/components/virtualRepeater/virtualRepeater.js

```diff
@@ -0,0 +1,452 @@
+  this.newStartIndex = 0;
+  this.newEndIndex = 0;
+  this.startIndex = 0;
+  this.endIndex = 0;
+  // TODO: measure width/height of first element from dom if not provided.
+  // getComputedStyle?
+  this.itemSize = $scope.$eval($attrs.mdSize);
+  this.blocks = {};
+  this.pooledBlocks = [];
+}
+
+
+/**
+ * Called at startup by the md-virtual-repeat postLink function.
+ * @param {!VirtualRepeatContainerController} The container's controller.
+ * @param {!Function} The repeated element's bound transclude function.
+ * @param {string} lhs The left hand side of the repeat expression, indicating
+ *     the name for each item in the array.
+ * @param {!Function} rhs A compiled expression based on the right hand side
+ *     of the repeat expression. Points to the array to repeat over.
+ */
+VirtualRepeatController.prototype.link_ = function(container, transclude, lhs, rhs, extraName) {
+  this.container = container;
+  this.transclude = transclude;
+  this.lhs = lhs;
+  this.rhs = rhs;
+  this.extraName = extraName;
+  this.sized = false;
+
+  this.container.register(this);
+};
+
+
+/**
+ * Called by the container. Informs us that the containers scroll or size has
+ * changed.
+ */
+VirtualRepeatController.prototype.containerUpdated = function() {
+  if (!this.sized) {
+    this.sized = true;
+    this.$scope.$watchCollection(this.rhs, angular.bind(this, this.virtualRepeatUpdate_));
+    this.items = this.rhs(this.$scope);
+  }
+
+  this.updateIndexes_();
+
+  if (this.newStartIndex !== this.startIndex ||
+      this.newEndIndex !== this.endIndex ||
+      this.container.getScrollOffset() > this.container.getScrollSize()) {
+    this.virtualRepeatUpdate_(this.items, this.items);
+  }
+};
+
+
+/**
+ * Called by the container. Returns the size of a single repeated item.
+ * @return {number} Size of a repeated item.
+ */
+VirtualRepeatController.prototype.getItemSize = function() {
+  return this.itemSize;
+};
+
+
+VirtualRepeatController.prototype.virtualRepeatUpdate_ = function(items, oldItems) {
+  var itemsLength = items ? items.length : 0;
+
+  // If the number if items shrank, scroll up to the top.
+  if (this.items && itemsLength < this.items.length && this.container.getScrollOffset() !== 0) {
+    this.items = items;
+    this.container.resetScroll();
+    return;
+  }
+
+  this.items = items;
+  if (items !== oldItems) {
+    this.updateIndexes_();
+  }
+
+  this.parentNode = this.$element[0].parentNode;
+  this.container.setScrollSize(itemsLength * this.itemSize);
+
+  // Detach and pool any blocks that are no longer in the viewport.
+  Object.keys(this.blocks).forEach(function(blockIndex) {
+    var index = parseInt(blockIndex, 10);
+    if (index < this.newStartIndex || index >= this.newEndIndex) {
+      this.poolBlock_(index);
+    }
+  }, this);
+
+  // Add needed blocks.
+  // For performance reasons, temporarily block browser url checks as we digest
+  // the restored block scopes.
+  this.$browser.$$checkUrlChange = angular.noop;
```

Comment:
Since `$$checkUrlChange` is not documented, perhaps it would be good to give a quick summary of what it does and why it's too slow for this.